ocf_datapipes/select/select_time_slice_nwp.py

```diff
@@ -78,23 +88,59 @@ def __iter__(self) -> Union[xr.DataArray, xr.Dataset]:
+            )
 
-            init_times = xr_available.sel(
+            # Find the most recent available init times for all target times
+            selected_init_times = available_init_times.sel(
                 init_time_utc=target_times,
                 method="ffill",  # forward fill from init times to target times
-            ).init_time_utc.values
+            ).values
 
-            steps = target_times - init_times
+            # Find the required steps for all target times
+            steps = target_times - selected_init_times
 
             # We want one timestep for each target_time_hourly (obviously!) If we simply do
             # nwp.sel(init_time=init_times, step=steps) then we'll get the *product* of
             # init_times and steps, which is not what # we want! Instead, we use xarray's
             # vectorized-indexing mode by using a DataArray indexer.  See the last example here:
             # https://docs.xarray.dev/en/latest/user-guide/indexing.html#more-advanced-indexing
             coords = {"target_time_utc": target_times}
-            init_time_indexer = xr.DataArray(init_times, coords=coords)
+            init_time_indexer = xr.DataArray(selected_init_times, coords=coords)
             step_indexer = xr.DataArray(steps, coords=coords)
-            xr_sel = xr_available.sel(step=step_indexer, init_time_utc=init_time_indexer)
+
+            # Slice out the data which does not need to be diffed
+            xr_non_accum = xr_data.sel({self.channel_dim_name: non_accum_channels})
+            xr_non_accum = xr_non_accum.sel(step=step_indexer, init_time_utc=init_time_indexer)
+
+            if len(accum_channels) == 0:
+                xr_sel = xr_non_accum
+
+            else:
+                # First minimise the size of the dataset we are diffing
+                # - find the init times we are slicing from
+                unique_init_times = np.unique(selected_init_times)
+                # - find the min and max steps we slice over. Max is extended due to diff
+                min_step = min(steps)
+                max_step = max(steps) + (xr_data.step[1] - xr_data.step[0])
+
+                xr_accum = xr_data.sel(
+                    {
+                        self.channel_dim_name: accum_channels,
+                        "init_time_utc": unique_init_times,
+                        "step": slice(min_step, max_step),
+                    }
+                )
+
+                # Take the diff and slice requested data
+                xr_accum = xr_accum.diff(dim="step", label="lower")
+                xr_sel_accum = xr_accum.sel(step=step_indexer, init_time_utc=init_time_indexer)
+
+                # Join diffed and non-diffed variables
+                xr_sel = xr.concat([xr_sel_non_accum, xr_sel_accum], dim=self.channel_dim_name)
+
+                # Reorder the variable back to the original order
+                xr_sel = xr_sel.sel({self.channel_dim_name: xr_data[self.channel_dim_name].values})
```

Comment:
Its it worth adding a warning, if the steps are not uniform? or something like that?

Reply:
That's was already a problem before these changes. The way is was/is currently written will cause an error if the steps are non-uniform. The uniform steps assumption also appears in more places than here. So I think that should just be made into a separate issue

Reply:
ok, then thats ok. thanks